### PR TITLE
feat: improve function autocomplete

### PR DESCRIPTION
**Pull request by scnwwu**
**Summary**
- Do NOT show functions as snippets
  - They're built-in syntax and should be distinguished from other snippets
  - Snippets will show everywhere which is sometimes unexpected for functions
- Provide functions autocomplete similar to other keywords from syntax data
  - so that it will be more accurate and up-to-date (reported by #457)
- Provide function autocomplete when typed characters don't match other keyword but functions

_Note_: Functions can be used in various places mixed with other keywords, it can't be 100% match user's expectation. User can specifically turn this feature off by VS Code settings `Editor › Suggest: Show Functions`

**Testing**

- Make sure functions autocomplete don't show as snippets
- Make sure functions autocomplete don't appear in invalid places like string, comment or cards data
- Make sure functions autocomplete can appear when typing matched characters (like sleep/rannor reported in #457)
- Make sure corresponding help description can be shown with the function autocomplete (click `>` on the list item)

### CODE REVIEW

**Review by Enzhihpp — APPROVED** [review]

File: server/src/sas/CompletionProvider.ts [hunk 1]
```diff
@@ -606,6 +628,39 @@ export class CompletionProvider {
               : `'${item.name.replace(/'/g, "''")}'n`,
         }));
 
+        if (
+          prefix.length > 1 &&
+          (items === undefined ||
+            items.length === 0 ||
+            !items.find((item) => item.label.startsWith(prefix))) &&
```
Comment (scottdover) [inline server/src/sas/CompletionProvider.ts:635]: It might be worth pulling out this condition into a variable that indicates what we're checking here

**Review by scottdover — APPROVED** [review]
Just had one comment around readability, but it isn't that big a deal

**Comment by Zhirong2022** [comment]
Acceptance Criteria:

1.Make sure there are no regression issues 
2.Make sure functions autocomplete don't show as snippets
3.Make sure functions autocomplete don't appear in invalid places like string, comment or cards data
4.Make sure functions autocomplete can appear when typing matched characters (like sleep/rannor reported in Missing Intellisense suggestions in LSP #457)
5.Make sure corresponding help description can be shown with the function autocomplete (click > on the list item)
6.Make sure functions autocomplete works well with VSC settings update
7.Verify 1 to 6 in OS: Windows and Mac
8.Verify 1 to 7 in NLS environment

**Comment by scnwwu** [comment]
> Function icon will be changed in certain case. Type code below and highlight 'intz' at line4, delete it and retype 'int' to show function autocomplete.
> 
> ```
> data _null_;
>     one      = 1;
>     int_one  = int(one);
>     intz_one = intz(one);
>  run;
>  
> ```
> 
> image

Fixed. Thank you

**Comment by scnwwu** [comment]
> No function autocomplete list if trying to update a function ends with semicolon. Type code below, double click %sysmexecdepth and delete. Type %sysmex to try to show the function autocomplete list.
> 
> ```
> %macro A;
>    %put %nrstr(%%)sysmexecdepth=%sysmexecdepth;
> %mend A;
> ```
> 
> image

Fixed. Thank you

**Comment by Zhirong2022** [comment]
The prompt is not correct for the nested function. Please put the code below, type ',' (comma) in %str() to show function prompt for %str().
```
%macro reptNote(month);
   %if %upcase(%str())) = DECEMBER %then
       %put NOTE: Year-end report;
   %else %put NOTE: Report for &month;
%mend reptNote;
```
image

**Comment by scnwwu** [comment]
> The prompt is not correct for the nested function. Please put the code below, type ',' (comma) in %str() to show function prompt for %str().
> 
> ```
> %macro reptNote(month);
>    %if %upcase(%str())) = DECEMBER %then
>        %put NOTE: Year-end report;
>    %else %put NOTE: Report for &month;
> %mend reptNote;
> ```
> 
> image

Fixed. Thank you

**Review by Zhirong2022 — APPROVED** [review]

**Comment by Zhirong2022** [comment]
All issues have been fixed.

**Pull request merged**